lms/models/lti_params.py

```diff
@@ -22,8 +24,47 @@ def v11(self):
         return self
 
     @classmethod
-    def from_v13(cls, v13_params):
-        return LTIParams(_to_lti_v11(v13_params), v13_params)
+    def from_request(cls, request):
+        """Create an LTIParams from the request."""
+
+        if v13_params := request.lti_jwt:
+            v11, v13 = _to_lti_v11(v13_params), v13_params
+        else:
+            v11, v13 = request.params, None
+
+        lti_params = cls(v11=v11, v13=v13)
+
+        # This would be good if we could extract this as a product plugin (or
+        # something similar), but unfortunately there's currently a circular
+        # dependency where LTI params are required to get the product
+        lti_params = _apply_canvas_quirks(lti_params, request)
+
+        return lti_params
+
+
+def _apply_canvas_quirks(lti_params, request):
+    # Canvas SpeedGrader launches LTI apps with the wrong resource_link_id,
+    # see:
+    #
+    # * https://github.com/instructure/canvas-lms/issues/1952
+    # * https://github.com/hypothesis/lms/issues/3228
+    #
+    # We add the correct resource_link_id as a query param on the launch
+    # URL that we submit to Canvas and use that instead of the incorrect
+    # resource_link_id that Canvas puts in the request's body.
+    is_speedgrader = request.params.get("learner_canvas_user_id")
+
+    if is_speedgrader and (resource_link_id := request.params.get("resource_link_id")):
```

Comment:
I think we want to be more explicit about where `resource_link_id` is coming from here. 

`request.GET` vs `request.params` (which squashes together GET and POST and I think in the case of LTI1.1 will always be there but it's not very clear which one we are picking/takes precedence for pyramid.).

I'd replace `request.params.get("resource_link_id")` by `request.GET.get("resource_link_id")`

Reply:
The same is true for `learner_canvas_user_id` but there's is less important as there's no clash between GET and POST params.